File: binarySearch.ts
const validateInputs = (numbers: number[], target: number) => {
  if (!Array.isArray(numbers) || !Number.isFinite(target)) {
    throw new Error("Inputs must be valid");

  }
}

const binarySearch = (numbers: number[], target: number) => {
  validateInputs(numbers, target)
  let left = 0
  let right = numbers.length - 1
  while (left <= right) {
    const mid = Math.floor((left + right) / 2)
    if (numbers[mid] === target) return mid
    else if (numbers[mid] < target) left = mid + 1
    else right = mid - 1
  }
  return -1
}

console.log(binarySearch([1, 2, 3, 49, 5], 49)) // 3
console.log(binarySearch([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 5)) // 4

File: binarySearch.py
from typing import Optional, Union
import math


def validate_numbers(numbers: list[float]):
    if not numbers:
        raise ValueError("List cannot be empty")


def binary_search(numbers: list[float], target: Union[int, float]) -> Optional[int]:
    validate_numbers(numbers)
    numbers = sorted(numbers)
    left = 0
    right = len(numbers) - 1
    while left <= right:
        mid = math.floor((left + right) / 2)
        if numbers[mid] == target:
            return mid
        elif numbers[mid] < target:
            left = mid + 1
        else:
            right = mid - 1
    return -1


print(binary_search([1, 2, 3, 4, 5, 6, 7, 8, 9, 10], 5))
